# Question 1.1: Write the Answer to these questions.

### 1. What is the difference between static and dynamic variables in Python

Static Variables

Definition: Static variables, also known as class variables, are shared across all instances of a 

class. They belong to the class itself rather than any particular object instance.

Scope: They are defined inside the class but outside any instance methods.

Usage: They are used when you want to maintain a single shared state across all instances of a class.

Dynamic Variables

Definition: Dynamic variables, also known as instance variables, are specific to each instance of a 

class. Each object has its own copy of an instance variable.

Scope: They are defined inside the constructor (usually __init__ method) or any other instance methods.

Usage: They are used to maintain unique state information for each object instance.

In [2]:
class MyClass:
    static_var = "I am static"

    def __init__(self, dynamic_var):
        self.dynamic_var = dynamic_var

print(MyClass.static_var)  

obj1 = MyClass("I am dynamic 1")
obj2 = MyClass("I am dynamic 2")

print(obj1.dynamic_var)  
print(obj2.dynamic_var)  

MyClass.static_var = "I am still static"

print(obj1.static_var) 
print(obj2.static_var)  


I am static
I am dynamic 1
I am dynamic 2
I am still static
I am still static


### 2. Explain the purpose of "pop","popitem","clear()" in a dictionary with suitable examples.

i. pop() : Removes and returns the value for the specified key. If the key is not found, a default value can be provided, otherwise, it raises a KeyError.
Syntax: dict.pop(key[, default])

In [5]:
my_dict = {'a': 1, 'b': 2, 'c': 3}
value = my_dict.pop('b')
print(value)       
print(my_dict)    

value = my_dict.pop('d', 'Not Found')
print(value)       

2
{'a': 1, 'c': 3}
Not Found


In [ ]:
ii. popitem() : Removes and returns the last inserted key-value pair as a tuple. Since Python 3.7, dictionaries maintain insertion order, so popitem() removes the last item inserted.
Syntax: dict.popitem()

In [6]:
my_dict = {'a': 1, 'b': 2, 'c': 3}
item = my_dict.popitem()
print(item)        
print(my_dict)   

('c', 3)
{'a': 1, 'b': 2}


In [ ]:
clear() : Removes all items from the dictionary, resulting in an empty dictionary.
Syntax: dict.clear()

In [7]:
my_dict = {'a': 1, 'b': 2, 'c': 3}
my_dict.clear()
print(my_dict)     

{}


### 3. What do you mean by FrozenSet? Explain it with suitable examples.

A frozenset in Python is an immutable version of a set. Unlike sets, which are mutable and can be modified (elements can be added or removed), frozensets are immutable and cannot be changed after their creation. This immutability makes frozensets hashable, meaning they can be used as keys in dictionaries or elements in other sets.

Characteristics of a FrozenSet

Immutable: Once created, the elements of a frozenset cannot be changed.
Hashable: Can be used as keys in dictionaries and elements of other sets.
Unordered: Elements do not have a specific order.
Unique Elements: Like sets, frozensets do not allow duplicate elements.

In [8]:
fset = frozenset([1, 2, 3, 4, 5])
print(fset)  

fset_from_set = frozenset({6, 7, 8, 9})
print(fset_from_set) 

frozenset({1, 2, 3, 4, 5})
frozenset({8, 9, 6, 7})


### 4. Differentiate between mutable and immutable data types in Python and give examples of mutable and immutable data types.

Modification Capability:
Mutable: Can be modified after creation.
Immutable: Cannot be modified after creation; any change creates a new object.

Examples:
Mutable: List, Dictionary, Set, Bytearray.
Immutable: String, Tuple, Integer, Float, Boolean, Frozenset, Bytes.

Memory Management:
Mutable: May remain in the same memory location or change it.
Immutable: Always results in a new memory location for any modification.

Performance:
Mutable: Generally faster for updates since modifications happen in place.
Immutable: Can be slower for updates as it involves creating new objects.

Use Cases:
Mutable: Suitable for scenarios where changes are expected, like data structures that grow or shrink dynamically.
Immutable: Ideal for constant values and hashable collections used as keys in dictionaries or elements in sets.

Examples of Use:
Mutable: Lists for collecting items, dictionaries for mapping key-value pairs, sets for unique elements.
Immutable: Strings for text data, tuples for fixed collections of items, integers and floats for numeric constants.

In [10]:
# Mutable
my_list = [1, 2, 3]
my_list[0] = 4  
print(my_list)  

my_dict = {'a': 1, 'b': 2}
my_dict['a'] = 3  
print(my_dict) 

my_set = {1, 2, 3}
my_set.add(4)  
print(my_set) 

[4, 2, 3]
{'a': 3, 'b': 2}
{1, 2, 3, 4}


In [11]:
# immutable
my_string = "hello"
new_string = my_string.replace('h', 'j') 
print(my_string)  
print(new_string) 

my_tuple = (1, 2, 3)
new_tuple = my_tuple + (4,)  
print(my_tuple)  
print(new_tuple) 

my_int = 5
new_int = my_int + 1  
print(my_int)  
print(new_int)  

hello
jello
(1, 2, 3)
(1, 2, 3, 4)
5
6


### 5. What is __init__?Explain with an example.

The __init__ method in Python is a special method called a constructor. It is automatically called when a new instance of a class is created. The primary purpose of __init__ is to initialize the instance's attributes with values provided as arguments.

In [13]:
class Person:
    def __init__(self, name, age):
        self.name = name  
        self.age = age  

    def greet(self):
        print(f"Hello, my name is {self.name} and I am {self.age} years old.")

person1 = Person("Alice", 30)
person2 = Person("Bob", 25)

print(person1.name) 
print(person2.age)   

person1.greet()  
person2.greet()  

Alice
25
Hello, my name is Alice and I am 30 years old.
Hello, my name is Bob and I am 25 years old.


### 6. What is docstring in Python?Explain with an example.

A docstring in Python is a special string literal that is used to document modules, classes, methods, and functions. It provides a convenient way to describe the purpose and usage of the code, and is accessible via the __doc__ attribute of the object it documents.

In [15]:
"""
This module demonstrates the use of docstrings.
It contains a class and a function with their own docstrings.
"""

class Calculator:
    """
    This class represents a simple calculator with basic arithmetic operations.
    """

    def __init__(self):
        """
        Initializes a new instance of the Calculator class.
        """
        pass

    def add(self, x, y):
        """
        Adds two numbers.

        Parameters:
        x (int or float): The first number.
        y (int or float): The second number.

        Returns:
        int or float: The sum of x and y.
        """
        return x + y

    def subtract(self, x, y):
        """
        Subtracts the second number from the first number.

        Parameters:
        x (int or float): The first number.
        y (int or float): The second number.

        Returns:
        int or float: The difference between x and y.
        """
        return x - y

print(Calculator.__doc__) 
print(Calculator.add.__doc__)  


    This class represents a simple calculator with basic arithmetic operations.
    

        Adds two numbers.

        Parameters:
        x (int or float): The first number.
        y (int or float): The second number.

        Returns:
        int or float: The sum of x and y.
        


### 7. What are unit tests in Python

Unit tests in Python are a type of testing used to verify that individual units or components of a program (such as functions or methods) work as expected. They are designed to test the smallest testable parts of an application in isolation from the rest of the codebase. The purpose of unit testing is to ensure that each unit of code performs correctly, making it easier to identify and fix bugs early in the development process.

### 8. What is break, continue and pass in Python


In [16]:
for i in range(10):
    if i == 5:
        break  
    print(i)

0
1
2
3
4


continue
Purpose: Skips the remaining code inside the current loop iteration and proceeds to the next iteration. It does not exit the loop.
Usage: Useful to skip certain iterations of the loop based on a condition.

In [17]:
for i in range(10):
    if i % 2 == 0:
        continue  
    print(i)

1
3
5
7
9


pass
Purpose: Acts as a placeholder that does nothing. It is used when a statement is required syntactically but you have no code to execute.
Usage: Useful in situations where code is required syntactically but you are not ready to implement the functionality.

In [18]:
for i in range(10):
    if i % 2 == 0:
        pass 
    else:
        print(i)

1
3
5
7
9


### 9. What is the use of self in Python?

Instance Reference: self refers to the current instance of the class, allowing access to its attributes and methods.

Mandatory in Method Definitions: It must be the first parameter in method definitions within a class, though it is not passed explicitly when the method is called.

Conventional Name: While self is the conventional name, it is not a keyword in Python; you can technically use any name, but using self is a widely accepted convention.

In [20]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age 

    def greet(self):
        print(f"Hello, my name is {self.name} and I am {self.age} years old.")

    def have_birthday(self):
        self.age += 1

person = Person("Alice", 30)

person.greet()       
person.have_birthday()  
person.greet()           

Hello, my name is Alice and I am 30 years old.
Hello, my name is Alice and I am 31 years old.


### 10. What are global, protected and private attributes in Python

Global Attributes
Definition: Global attributes are variables that are defined outside of any class or function and are accessible from anywhere within the module.

Usage: They are used for data that needs to be shared across multiple classes or functions within the same module.Global Attributes
Definition: Global attributes are variables that are defined outside of any class or function and are accessible from anywhere within the module.

Usage: They are used for data that needs to be shared across multiple classes or functions within the same module.

In [21]:
global_variable = "I am global"

def display_global():
    print(global_variable)

class MyClass:
    def show_global(self):
        print(global_variable)

display_global() 
obj = MyClass()
obj.show_global()  

I am global
I am global


Protected Attributes

Definition: Protected attributes are meant to be accessed only within the class and its subclasses. They are not intended to be accessed from outside the class hierarchy.

Naming Convention: They are indicated by a single leading underscore (_).

Usage: To suggest that these attributes should not be accessed directly from outside the class, but it is not enforced by the language.

In [22]:
class Parent:
    def __init__(self):
        self._protected_attr = "I am protected"

class Child(Parent):
    def show_protected(self):
        print(self._protected_attr)

obj = Child()
obj.show_protected()  

print(obj._protected_attr) 

I am protected
I am protected


Private Attributes

Definition: Private attributes are meant to be accessible only within the class where they are defined. They are not accessible from outside the class or from subclasses.

Naming Convention: They are indicated by a double leading underscore (__).

Usage: To enforce encapsulation and restrict access to the attribute from outside the class.

In [24]:
class MyClass:
    def __init__(self):
        self.__private_attr = "I am private"

    def show_private(self):
        print(self.__private_attr)

obj = MyClass()
obj.show_private() 

try:
    print(obj.__private_attr)
except AttributeError as e:
    print(e) 

print(obj._MyClass__private_attr)  

I am private
'MyClass' object has no attribute '__private_attr'
I am private


### 11. What are modules and packages in Python

Modules
Definition: A module is a single file containing Python code. It can include functions, classes, variables, and runnable code. Modules help in organizing code into logical units.

Usage:
Code Organization: Helps in separating code into different files based on functionality.
Reusability: Allows reuse of code across different programs or parts of a program.

Packages
Definition: A package is a directory containing multiple modules and an __init__.py file. The __init__.py file can be empty but must be present to indicate that the directory is a package. Packages help in organizing related modules into a directory hierarchy.

Usage:
Hierarchical Organization: Allows grouping of related modules into a directory structure.
Namespace Management: Helps in managing namespaces and avoiding module name conflicts.

### 12. What are lists and tuples? What is the key difference between the two?

Lists
Definition: Lists are mutable sequences in Python. They can contain items of any type and allow modification of their elements.

Key Features:
Mutable: You can change, add, or remove items after the list has been created.
Dynamic Size: Lists can grow or shrink in size.
Syntax: Defined using square brackets [].

In [25]:
my_list = [1, 2, 3, "hello", 4.5]
my_list.append(6)       
my_list[0] = 10          
my_list.remove("hello")  

print(my_list) 

[10, 2, 3, 4.5, 6]


Tuples
Definition: Tuples are immutable sequences in Python. Once a tuple is created, its contents cannot be changed.

Key Features:

Immutable: You cannot modify, add, or remove items once the tuple has been created.
Fixed Size: The size of a tuple is fixed once created.
Syntax: Defined using parentheses ().

In [26]:
my_tuple = (1, 2, 3, "hello", 4.5)

print(my_tuple)        

(1, 2, 3, 'hello', 4.5)


### 13. What is an Interpreted language & dynamically typed language?Write 5 differences between them?

Interpreted Language

Definition: An interpreted language is a type of programming language for which most of the instructions are executed directly by an interpreter, rather than being compiled into machine code. The interpreter reads and executes the code line-by-line or statement-by-statement.

Dynamically Typed Language

Definition: A dynamically typed language is one where the type of a variable is determined at runtime, rather than at compile-time. This means that you do not need to declare the type of a variable explicitly.

Difference Between Interpreted and Dynamically Typed Languages
Conceptual Focus:

Interpreted Language: Focuses on how code is executed (line-by-line by an interpreter).
Dynamically Typed Language: Focuses on how types are handled (types are determined at runtime).
Execution:

Interpreted Language: Code is executed directly by the interpreter, which can lead to slower execution compared to compiled languages.
Dynamically Typed Language: The code execution might be interpreted or compiled, but type information is managed at runtime.
Type Checking:

Interpreted Language: Type checking may be performed by the interpreter if the language is also dynamically typed.
Dynamically Typed Language: Type checking is always done at runtime, regardless of whether the language is interpreted or compiled.
Error Handling:

Interpreted Language: Errors are typically detected during execution, which can include syntax errors or runtime errors.
Dynamically Typed Language: Type-related errors are detected at runtime, and other errors depend on whether the language is interpreted or compiled.
Development and Debugging:

Interpreted Language: Allows for rapid development and testing, as changes can be tested immediately without recompiling.
Dynamically Typed Language: Provides flexibility in coding and allows for dynamic variable assignments, which can sometimes lead to runtime type errors.

### 14. What are Dict and List comprehensions?

List Comprehensions
Definition: List comprehensions provide a way to create lists by specifying an expression and an iteration over an iterable. They can include optional conditions to filter items.

In [28]:
evens = [x for x in range(10) if x % 2 == 0]
print(evens)  

[0, 2, 4, 6, 8]


Dict Comprehensions
Definition: Dict comprehensions provide a way to create dictionaries by specifying key-value pairs in a similar way to list comprehensions. They also allow for conditions to be included.

In [29]:
even_squares_dict = {x: x**2 for x in range(10) if x % 2 == 0}
print(even_squares_dict) 

{0: 0, 2: 4, 4: 16, 6: 36, 8: 64}


### 15. What are decorators in Python? Explain it with an example.Write down its use cases.

Decorators: 
Definition: A decorator is a function that takes another function (or method) as its argument, adds some functionality, and returns a new function with the additional behavior.

In [30]:
def my_decorator(func):
    def wrapper(*args, **kwargs):
        print("Something is happening before the function is called.")
        result = func(*args, **kwargs)
        print("Something is happening after the function is called.")
        return result
    return wrapper

In [31]:
@my_decorator
def say_hello(name):
    print(f"Hello, {name}!")

say_hello("Alice")

Something is happening before the function is called.
Hello, Alice!
Something is happening after the function is called.


Logging:
Purpose: To add logging functionality to functions, such as logging the execution time or parameters.

Authorization:
Purpose: To control access to certain functions based on user roles or permissions.

Caching:
Purpose: To cache results of expensive function calls and reuse them to improve performance.

Validation:
Purpose: To validate input arguments of a function before proceeding with its execution.

Timing:
Purpose: To measure and report the time taken by a function to execute.

### 16. How is memory managed in Python

Memory management
Reference Counting:
Definition: Each object in Python has an associated reference count that keeps track of how many references point to that object.

Garbage Collection
Definition: Garbage collection is a mechanism to detect and clean up memory that is no longer in use but not yet freed due to circular references.

Memory Allocation
Private Heap Space: Python maintains an internal private heap space where all the objects and data structures are stored.

Memory Management for Built-in Types
Immutable Types: Immutable objects like integers, floats, and strings are managed differently. Python reuses small immutable objects to save memory (e.g., small integers are cached and reused).

Mutable Types: Mutable objects like lists and dictionaries have different management strategies to handle resizing and reallocation.

In [32]:
import sys

a = 1000
b = 1000
print(sys.getrefcount(a))  

list1 = [1, 2, 3]
list2 = list1
print(sys.getrefcount(list1))  

3
3


### 17. What is lambda in Python? Why is it used

Lambda functions are often used for the following reasons:

Conciseness:
They provide a way to create small, one-off functions without having to formally define them using the def keyword. This can make the code more concise and readable when the function is simple.

Usage: They are used for their conciseness, particularly in functional programming scenarios, custom sorting, event handling, or wherever a simple function is needed without a formal definition.

In [33]:
add = lambda x, y: x + y

result = add(5, 3)
print(result) 

8


### 18. Explain split() and join() functions in Python.

split()
Definition: The split() method is used to divide a string into a list of substrings based on a specified delimiter or separator. By default, it splits the string by whitespace.

In [35]:
text = "one two three four"
parts = text.split(' ', 2)
print(parts)  

['one', 'two', 'three four']


# join()
Definition: The join() method is used to concatenate a list of strings into a single string with a specified separator or delimiter between each element.

In [36]:
numbers = ['1', '2', '3']
text = '-'.join(numbers)
print(text) 

1-2-3


### 19. What are iterators , iterable & generators in Python

Iterables
Definition: An iterable is any Python object capable of returning its members one at a time, allowing it to be iterated over in a for-loop or other iteration contexts.

In [37]:
numbers = [1, 2, 3, 4, 5]

for number in numbers:
    print(number)

1
2
3
4
5


Iterators
Definition: An iterator is an object that represents a stream of data. It implements the __iter__() and __next__() methods, allowing it to iterate over its data one element at a time.

In [38]:
numbers = [1, 2, 3]
iterator = iter(numbers)

print(next(iterator))  
print(next(iterator))  
print(next(iterator))  

1
2
3


Generators
Definition: Generators are a special type of iterator that are defined using functions with the yield keyword. They provide a way to create iterators in a more concise manner. 


In [41]:
def generate_numbers(n):
    for i in range(n):
        yield i

gen = generate_numbers(5)

for number in gen:
    print(number)

0
1
2
3
4


### 20. What is the difference between xrange and range in Python.

range:

Definition: Returns a list containing all the numbers in the specified range.
Memory Usage: Generates and stores all the numbers in memory as a list, which can be inefficient for large ranges.

In [42]:
numbers = range(5)
print(numbers)  

range(0, 5)


xrange:

Definition: Returns an iterator that generates numbers on-the-fly as you iterate over it.
Memory Usage: More memory-efficient as it does not store the entire sequence in memory, only generating numbers as needed.

In [43]:
numbers = range(5)
print(numbers)

range(0, 5)


### 21. Pillars of Oops8

1. Encapsulation
Definition: Encapsulation is the concept of bundling data (attributes) and methods (functions) that operate on the data into a single unit called a class. It also involves restricting direct access to some of an object's components, which can help in protecting the internal state of the object.

Benefits:

Data Hiding: Keeps the internal state of an object hidden from the outside world and only exposes a controlled interface.
Modularity: Promotes modularity by grouping related data and methods together.

In [44]:
class Car:
    def __init__(self, make, model):
        self.__make = make  
        self.__model = model 

    def get_make(self):
        return self.__make

    def set_make(self, make):
        self.__make = make

my_car = Car("Toyota", "Corolla")
print(my_car.get_make())  
my_car.set_make("Honda")

Toyota


2. Inheritance
Definition: Inheritance is a mechanism by which one class (the child or subclass) inherits attributes and methods from another class (the parent or superclass). This promotes code reusability and establishes a hierarchical relationship between classes.

Types:

Single Inheritance: A class inherits from a single parent class.
Multiple Inheritance: A class inherits from more than one parent class.
Multilevel Inheritance: A class inherits from another class, which is also inherited by a third class.
Hierarchical Inheritance: Multiple classes inherit from a single parent class.

In [45]:
class Animal:
    def speak(self):
        print("Animal speaks")

class Dog(Animal):
    def bark(self):
        print("Dog barks")

my_dog = Dog()
my_dog.speak()  
my_dog.bark()  

Animal speaks
Dog barks


3. Polymorphism
Definition: Polymorphism means "many shapes" and refers to the ability of different objects to respond, each in its own way, to the same method call. It allows for a single interface to be used for a general class of actions, with the specific action determined by the exact nature of the situation.

Types:

Method Overloading: Defining multiple methods with the same name but different parameters within the same class (not natively supported in Python but can be emulated).
Method Overriding: Subclasses providing a specific implementation of a method that is already defined in its superclass.

In [46]:
class Bird:
    def speak(self):
        print("Bird chirps")

class Dog:
    def speak(self):
        print("Dog barks")

def make_it_speak(animal):
    animal.speak()

my_bird = Bird()
my_dog = Dog()

make_it_speak(my_bird) 
make_it_speak(my_dog)   

Bird chirps
Dog barks


4. Abstraction
Definition: Abstraction is the concept of hiding the complex implementation details and showing only the necessary features of an object. It allows focusing on what an object does rather than how it does it.

Implementation:

Abstract Classes: Classes that cannot be instantiated and are meant to be subclasses. They often contain one or more abstract methods that must be implemented by subclasses.
Abstract Methods: Methods that are declared in an abstract class but contain no implementation. Subclasses must provide an implementation for these methods.

In [47]:
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

my_rectangle = Rectangle(5, 10)
print(my_rectangle.area()) 

50


### 22. How will you check if a class is a child of another class

issubclass() Function
subclass: The class you want to check.
superclass: The class you are checking against.
Returns:

True if subclass is a subclass of superclass.
False otherwise.

In [48]:
class Animal:
    pass

class Mammal(Animal):
    pass

class Dog(Mammal):
    pass

print(issubclass(Dog, Mammal))  
print(issubclass(Dog, Animal)) 
print(issubclass(Mammal, Dog)) 

True
True
False


### 23.How does inheritance work in python? Explain all types of inheritance with an example

How Inheritance Works in Python
Basic Concept:

Parent Class (Base Class): The class from which attributes and methods are inherited.
Child Class (Derived Class): The class that inherits from the parent class. It can extend or modify the behavior of the parent class.

1. Single Inheritance
Definition: A child class inherits from one parent class.

In [51]:
class Animal:
    def eat(self):
        print("Animal is eating")

class Dog(Animal):
    def bark(self):
        print("Dog is barking")

my_dog = Dog()
my_dog.eat()  
my_dog.bark()  

Animal is eating
Dog is barking


4.Hierarchical Inheritance
Definition: Multiple child classes inherit from a single parent class.

In [52]:
class Animal:
    def eat(self):
        print("Animal is eating")

class Dog(Animal):
    def bark(self):
        print("Dog is barking")

class Cat(Animal):
    def meow(self):
        print("Cat is meowing")

my_dog = Dog()
my_cat = Cat()
my_dog.eat()  
my_dog.bark() 
my_cat.eat()  
my_cat.meow() 

Animal is eating
Dog is barking
Animal is eating
Cat is meowing


In [53]:
class A:
    def method_a(self):
        print("Method A")

class B(A):
    def method_b(self):
        print("Method B")

class C(A):
    def method_c(self):
        print("Method C")

class D(B, C):
    pass

my_d = D()
my_d.method_a() 
my_d.method_b() 
my_d.method_c()  

Method A
Method B
Method C


### 24. What is encapsulation? Explain it with an example.

In [54]:
class BankAccount:
    def __init__(self, owner, balance=0):
        self.owner = owner 
        self.__balance = balance  

    def get_balance(self):
        return self.__balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount
            print(f"Deposited ${amount}. New balance is ${self.__balance}.")
        else:
            print("Deposit amount must be positive.")

    def withdraw(self, amount):
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            print(f"Withdrew ${amount}. New balance is ${self.__balance}.")
        else:
            print("Invalid withdrawal amount or insufficient funds.")

account = BankAccount("John Doe", 1000)

print(account.get_balance())  
account.deposit(500)          
account.withdraw(200)      

print(account.get_balance())  

1000
Deposited $500. New balance is $1500.
Withdrew $200. New balance is $1300.
1300


### 25.What is polymorphism? Explain it with an example.

In [55]:
class Animal:
    def speak(self):
        print("Animal speaks")

class Dog(Animal):
    def speak(self):
        print("Dog barks")

class Cat(Animal):
    def speak(self):
        print("Cat meows")

def make_animal_speak(animal):
    animal.speak()

my_dog = Dog()
my_cat = Cat()

make_animal_speak(my_dog) 
make_animal_speak(my_cat) 

Dog barks
Cat meows


### Question 1. 2. Which of the following identifier names are invalid and why? 

Evaluation of Each Identifier
a) Serial_no.
Invalid: Ends with a period (.), which is not allowed in identifiers.

b) 1st_Room
Invalid: Starts with a digit (1), which is not allowed. Identifiers must start with a letter or an underscore.

c) Hundred$
Invalid: Contains a dollar sign ($), which is not allowed in identifiers.

d) Total_Marks
Valid: Follows all the rules for identifiers. Contains only letters, digits (not at the start), and underscores.

e) total-Marks
Invalid: Contains a hyphen (-), which is not allowed in identifiers.

f) Total Marks
Invalid: Contains a space, which is not allowed in identifiers.

g) True
Invalid: True is a reserved keyword in Python and cannot be used as an identifier.

h) _Percentag
Valid: Follows all the rules for identifiers. Starts with an underscore and contains only letters and an underscore.

### Question 1.3

name = ["Mohan", "dash", "karam", "chandra","gandhi","Bapu"]
do the following operations in this list;

a) add an element "freedom_fighter" in this list at the 0th index.

In [57]:
name = ["Mohan", "dash", "karam", "chandra", "gandhi", "Bapu"]
name.insert(0, "freedom_fighter")
print(name)

['freedom_fighter', 'Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapu']


In [59]:
b) find the output of the following ,and explain how?

name = ['freedomFighter',"Bapuji","MOhan" "dash", "karam",
"chandra","gandhi"]
length1=len((name[-len(name)+1:-1:2]))
length2=len((name[-len(name)+1:-1]))
print(length1+length2)

Object `how` not found.
6


c) add two more elements in the name ["NetaJi","Bose"] at the end of the list.

In [60]:
name = ["freedomFighter", "Bapuji", "MOhan", "dash", "karam", "chandra", "gandhi"]

name.append("NetaJi")
name.append("Bose")

print(name)

['freedomFighter', 'Bapuji', 'MOhan', 'dash', 'karam', 'chandra', 'gandhi', 'NetaJi', 'Bose']


d) what will be the value of temp:
name = ["Bapuji", "dash", "karam", "chandra","gandi","Mohan"]
temp=name[-1]
name[-1]=name[0]
name[0]=temp
print(name)

In [61]:
name = ["Bapuji", "dash", "karam", "chandra", "gandi", "Mohan"]
temp = name[-1]  
name[-1] = name[0]  
name[0] = temp  
print(name)  

['Mohan', 'dash', 'karam', 'chandra', 'gandi', 'Bapuji']


### Question 1.4.Find the output of the following.

In [63]:
animal = ['Human','cat','mat','cat','rat','Human', 'Lion']
print(animal.count('Human'))
print(animal.index('rat'))
print(len(animal))

2
4
7


### 1.5 tuple1=(10,20,"Apple",3.4,'a',["master","ji"],("sita","geeta",22),[{"roll_no"N1}, {"name"N"Navneet"}])

a)print(len(tuple))

In [64]:
tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no": "N1"}, {"name": "Navneet"}])
print(len(tuple1))


8


b) print(tuple[-1][-1]['name'])

In [65]:
tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no": "N1"}, {"name": "Navneet"}])
print(tuple1[-1][-1]['name'])

Navneet


c)fetch the value of roll_no from this tuple.

In [66]:
tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no": "N1"}, {"name": "Navneet"}])
roll_no_value = tuple1[-1][0]['roll_no']

print(roll_no_value)

N1


d)print(tuple1[-3][1])

In [67]:
tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no": "N1"}, {"name": "Navneet"}])
tuple1[-3][1]

'ji'

e)fetch the element "22" from this tuple.

In [68]:
tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no": "N1"}, {"name": "Navneet"}])
element_22 = tuple1[-2][2]


print(element_22)

22


### 1.6. Write a program to display the appropriate message as per the color of signal(RED-Stop/Yellow-Stay/ Green-Go) at the road crossing

In [69]:
def traffic_signal_message(color):
    if color == "RED":
        return "Stop"
    elif color == "YELLOW":
        return "Stay"
    elif color == "GREEN":
        return "Go"
    else:
        return "Invalid color"

signal_color = input("Enter the traffic signal color (RED, YELLOW, GREEN): ").strip().upper()

message = traffic_signal_message(signal_color)
print(message)

Enter the traffic signal color (RED, YELLOW, GREEN): grren
Invalid color


### 1.7. Write a program to create a simple calculator performing only four basic operations(+,-,/,*) .

In [70]:
def calculate(num1, num2, operation):
    if operation == '+':
        return num1 + num2
    elif operation == '-':
        return num1 - num2
    elif operation == '*':
        return num1 * num2
    elif operation == '/':
        if num2 == 0:
            return "Error: Division by zero is not allowed."
        return num1 / num2
    else:
        return "Error: Invalid operation."

def main():
    print("Simple Calculator")
    print("Available operations: +, -, *, /")
    
    try:
        num1 = float(input("Enter the first number: "))
    except ValueError:
        print("Invalid input! Please enter a number.")
        return
    
    try:
        num2 = float(input("Enter the second number: "))
    except ValueError:
        print("Invalid input! Please enter a number.")
        return
    
    operation = input("Enter the operation (+, -, *, /): ").strip()
    
    result = calculate(num1, num2, operation)
    print(f"Result: {result}")

if __name__ == "__main__":
    main()

Simple Calculator
Available operations: +, -, *, /
Enter the first number: 45
Enter the second number: 12
Enter the operation (+, -, *, /): +
Result: 57.0


### 1.8. Write a program to find the larger of the three pre-specified numbers using ternary operators.

In [71]:
num1 = 15
num2 = 25
num3 = 10

largest = num1 if (num1 >= num2 and num1 >= num3) else (num2 if num2 >= num3 else num3)

print("The largest number is:", largest)

The largest number is: 25


### 1.9. Write a program to find the factors of a whole number using a while loop. 

In [72]:
def find_factors(number):
    if number <= 0:
        print("Please enter a positive whole number.")
        return
    
    i = 1
    print(f"Factors of {number} are:")
    
    while i <= number:
        if number % i == 0:
            print(i, end=' ')
        i += 1

try:
    num = int(input("Enter a whole number: "))
    find_factors(num)
except ValueError:
    print("Invalid input! Please enter a whole number.")

Enter a whole number: 12
Factors of 12 are:
1 2 3 4 6 12 

### 1.10. Write a program to find the sum of all the positive numbers entered by the user. As soon as the user 

In [73]:
def sum_positive_numbers():
    total_sum = 0
    
    while True:
        try:
            number = float(input("Enter a positive number (or a negative number to stop): "))
            
            if number < 0:
                break
            
            total_sum += number
        
        except ValueError:
            print("Invalid input! Please enter a valid number.")
    
    print(f"The sum of all positive numbers entered is: {total_sum}")

sum_positive_numbers()

Enter a positive number (or a negative number to stop): 25
Enter a positive number (or a negative number to stop): -31
The sum of all positive numbers entered is: 25.0


### 1.11. Write a program to find prime numbers between 2 to 100 using nested for loops.

In [74]:
def is_prime(n):
    """Return True if n is a prime number, otherwise False."""
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

def find_primes_in_range(start, end):
    """Print all prime numbers in the range [start, end]."""
    print(f"Prime numbers between {start} and {end} are:")
    for num in range(start, end + 1):
        if is_prime(num):
            print(num, end=' ')

find_primes_in_range(2, 100)

Prime numbers between 2 and 100 are:
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 

## 1.12 Write the programs for the following
i. Accept the marks of the student in five major subjects and display the same.

Grade A: Percentage > 85
Grade B: 85 < && Percentage >= 75
Grade C: 75 < && Percentage >= 50
Grade D: 50 <= Percentage > 30
Reappear: Percentage < 30

In [75]:
def calculate_grade(percentage):
    if percentage > 85:
        return "A"
    elif percentage <= 85 and percentage > 75:
        return "B"
    elif percentage <= 75 and percentage > 50:
        return "C"
    elif percentage <= 50 and percentage > 30:
        return "D"
    else:
        return "Reappear"

def calculate_percentage(marks):
    return sum(marks) / len(marks)

def main():
    subjects = ['Subject 1', 'Subject 2', 'Subject 3', 'Subject 4', 'Subject 5']
    marks = []

    for subject in subjects:
        mark = float(input(f"Enter marks for {subject}: "))
        marks.append(mark)

    percentage = calculate_percentage(marks)
    grade = calculate_grade(percentage)

    print("\nMarks Obtained:")
    for i, mark in enumerate(marks):
        print(f"{subjects[i]}: {mark}")

    print(f"\nPercentage: {percentage}%")
    print(f"Grade: {grade}")

if __name__ == "__main__":
    main()

Enter marks for Subject 1: 58
Enter marks for Subject 2: 88
Enter marks for Subject 3: 65
Enter marks for Subject 4: 75
Enter marks for Subject 5: 65

Marks Obtained:
Subject 1: 58.0
Subject 2: 88.0
Subject 3: 65.0
Subject 4: 75.0
Subject 5: 65.0

Percentage: 70.2%
Grade: C


### ii )Calculate the sum of the marks of all subjects. Divide the total marks by number of subjects (i.e. 5), calculate percentage = total marks/5 and display the percentage

In [76]:
def calculate_grade(percentage):
    if percentage > 85:
        return "A"
    elif percentage <= 85 and percentage > 75:
        return "B"
    elif percentage <= 75 and percentage > 50:
        return "C"
    elif percentage <= 50 and percentage > 30:
        return "D"
    else:
        return "Reappear"

def main():
    subjects = ['Subject 1', 'Subject 2', 'Subject 3', 'Subject 4', 'Subject 5']
    marks = []

    for subject in subjects:
        mark = float(input(f"Enter marks for {subject}: "))
        marks.append(mark)

    total_marks = sum(marks)
    percentage = total_marks / len(subjects)
    grade = calculate_grade(percentage)

    print("\nMarks Obtained:")
    for i, mark in enumerate(marks):
        print(f"{subjects[i]}: {mark}")

    print(f"\nTotal Marks: {total_marks}")
    print(f"Percentage: {percentage}%")
    print(f"Grade: {grade}")

if __name__ == "__main__":
    main()

Enter marks for Subject 1: 56
Enter marks for Subject 2: 86
Enter marks for Subject 3: 89
Enter marks for Subject 4: 78
Enter marks for Subject 5: 74

Marks Obtained:
Subject 1: 56.0
Subject 2: 86.0
Subject 3: 89.0
Subject 4: 78.0
Subject 5: 74.0

Total Marks: 383.0
Percentage: 76.6%
Grade: B


### iii) Find the grade of the student as Ter the following criteria . Hint: Use Match & case for this.: 

In [77]:
def calculate_grade(percentage):
    match percentage:
        case _ if percentage > 85:
            return "A"
        case _ if 75 < percentage <= 85:
            return "B"
        case _ if 50 < percentage <= 75:
            return "C"
        case _ if 30 < percentage <= 50:
            return "D"
        case _ if percentage <= 30:
            return "Reappear"

def main():
    subjects = ['Subject 1', 'Subject 2', 'Subject 3', 'Subject 4', 'Subject 5']
    marks = []

    for subject in subjects:
        mark = float(input(f"Enter marks for {subject}: "))
        marks.append(mark)

    total_marks = sum(marks)
    percentage = total_marks / len(subjects)
    grade = calculate_grade(percentage)

    print("\nMarks Obtained:")
    for i, mark in enumerate(marks):
        print(f"{subjects[i]}: {mark}")

    print(f"\nTotal Marks: {total_marks}")
    print(f"Percentage: {percentage}%")
    print(f"Grade: {grade}")

if __name__ == "__main__":
    main()

Enter marks for Subject 1: 89
Enter marks for Subject 2: 57
Enter marks for Subject 3: 36
Enter marks for Subject 4: 48
Enter marks for Subject 5: 68

Marks Obtained:
Subject 1: 89.0
Subject 2: 57.0
Subject 3: 36.0
Subject 4: 48.0
Subject 5: 68.0

Total Marks: 298.0
Percentage: 59.6%
Grade: C


### 1.13. Write a program for VIBGYOR Spectrum based on their Wavelength using. Wavelength Range:

In [ ]:
Color	Wavelength (nm)
Violet	400.0 - 440.0
Indigo	440.0 - 460.0
Blue	460.0 - 500.0
Green	500.0 - 570.0
Yellow	570.0 - 590.0
Orange	590.0 - 620.0
Red	620.0 - 720.0

In [78]:
def determine_color(wavelength):
    match wavelength:
        case _ if 400.0 <= wavelength < 440.0:
            return "Violet"
        case _ if 440.0 <= wavelength < 460.0:
            return "Indigo"
        case _ if 460.0 <= wavelength < 500.0:
            return "Blue"
        case _ if 500.0 <= wavelength < 570.0:
            return "Green"
        case _ if 570.0 <= wavelength < 590.0:
            return "Yellow"
        case _ if 590.0 <= wavelength < 620.0:
            return "Orange"
        case _ if 620.0 <= wavelength <= 720.0:
            return "Red"
        case _:
            return "Wavelength out of visible spectrum"

def main():
    wavelength = float(input("Enter the wavelength (in nm): "))
    color = determine_color(wavelength)
    print(f"The color corresponding to the wavelength {wavelength} nm is: {color}")

if __name__ == "__main__":
    main()

Enter the wavelength (in nm): 566
The color corresponding to the wavelength 566.0 nm is: Green


### 1.14.Consider the gravitational interactions between the Earth, Moon, and Sun in our solar system.

In [ ]:
mass_earth = 5.972e24 # Mass of Earth in kilograms
mass_moon = 7.34767309e22 # Mass of Moon in kilograms
mass_sun = .989e30 # Mass of Sun in kilograms
distance_earth_sun = .496e # Average distan0e between Earth and Sun in meters
distance_moon_earth = 3.844e8 # Average distan0e between Moon and Earth in meters

### i) Calculate the gravitational for0e between the Earth and the Sun 

In [79]:
def gravitational_force(m1, m2, r):
    G = 6.67430e-11  # Gravitational constant in m^3 kg^-1 s^-2
    return G * m1 * m2 / r**2

mass_earth = 5.972e24  # Mass of Earth in kg
mass_sun = 0.989e30    # Mass of Sun in kg
distance_earth_sun = 1.496e11  # Distance between Earth and Sun in meters

force_earth_sun = gravitational_force(mass_earth, mass_sun, distance_earth_sun)
print(f"The gravitational force between the Earth and the Sun is approximately {force_earth_sun:.2e} N.")

The gravitational force between the Earth and the Sun is approximately 1.76e+22 N.


### ii) Calculate the gravitational for0e between the Moon and the Earth.

In [80]:
def gravitational_force(m1, m2, r):
    G = 6.67430e-11  # Gravitational constant in m^3 kg^-1 s^-2
    return G * m1 * m2 / r**2

mass_earth = 5.972e24  # Mass of Earth in kg
mass_moon = 7.34767309e22  # Mass of Moon in kg
distance_moon_earth = 3.844e8  # Distance between Moon and Earth in meters

force_moon_earth = gravitational_force(mass_earth, mass_moon, distance_moon_earth)
print(f"The gravitational force between the Moon and the Earth is approximately {force_moon_earth:.2e} N.")

The gravitational force between the Moon and the Earth is approximately 1.98e+20 N.


### iii) Compare the calculated forces to determine which gravitational force is stronger. 

In [81]:
def gravitational_force(m1, m2, r):
    G = 6.67430e-11  # Gravitational constant in m^3 kg^-1 s^-2
    return G * m1 * m2 / r**2

# Given values
mass_earth = 5.972e24  # Mass of Earth in kg
mass_moon = 7.34767309e22  # Mass of Moon in kg
mass_sun = 0.989e30    # Mass of Sun in kg
distance_earth_sun = 1.496e11  # Distance between Earth and Sun in meters
distance_moon_earth = 3.844e8  # Distance between Moon and Earth in meters

# Calculate gravitational forces
force_earth_sun = gravitational_force(mass_earth, mass_sun, distance_earth_sun)
force_moon_earth = gravitational_force(mass_earth, mass_moon, distance_moon_earth)

# Compare the forces
stronger_force = "Sun-Earth" if force_earth_sun > force_moon_earth else "Moon-Earth"

print(f"Gravitational force between the Earth and the Sun: {force_earth_sun:.2e} N")
print(f"Gravitational force between the Earth and the Moon: {force_moon_earth:.2e} N")
print(f"The stronger gravitational force is between the {stronger_force}.")

Gravitational force between the Earth and the Sun: 1.76e+22 N
Gravitational force between the Earth and the Moon: 1.98e+20 N
The stronger gravitational force is between the Sun-Earth.


### iv) Explain which celestial body (Earth or Moon is more attracted to the other based on the comparison.

In the context of gravitational attraction, Newton's Third Law states that the force exerted by one object on another is equal in magnitude and opposite in direction. This means that the Earth and the Moon exert the same amount of gravitational force on each other. The same principle applies to the Earth and the Sun.

Equal Forces: The gravitational force the Earth exerts on the Moon is exactly equal to the gravitational force the Moon exerts 
on the Earth. The same is true for the Earth-Sun system.

Effect on Motion: Even though the forces are equal, the effect of these forces on each body depends on their masses. The less massive object (Moon) experiences a greater acceleration due to the force compared to the more massive object (Earth).

Attraction Implications:
Earth and Moon: Despite the Moon's much smaller mass compared to the Earth, they mutually attract each other with the same force. However, the Moon's smaller mass means it accelerates more towards the Earth than the Earth does towards the Moon.

Earth and Sun: The same principle applies between the Earth and the Sun. The Earth is much less massive than the Sun, so the Earth accelerates more towards the Sun than the Sun does towards the Earth, despite the forces being equal.

Conclusion:
Attraction to Earth or Moon: Both the Earth and the Moon are equally attracted to each other, but the Moon experiences a greater acceleration due to its smaller mass.

Attraction to Earth or Sun: Similarly, both the Earth and the Sun are equally attracted to each other, but the Earth experiences a greater acceleration due to its much smaller mass compared to the Sun.

2. Design and implement a Python program for managing student information using object-oriented
principles. Create a class called `Student` with encapsulated attributes for name, age, and roll number.
Implement getter and setter methods for these attributes. Additionally, provide methods to display student
information and update student details.

In [1]:
class Student:
    def __init__(self, name, age, roll_number):
        self.__name = name
        self.__age = age
        self.__roll_number = roll_number
    def get_name(self):
        return self.__name    
    def set_name(self, name):
        self.__name = name
    def get_age(self):
        return self.__age
    def set_age(self, age):
        self.__age = age
    def get_roll_number(self):
        return self.__roll_number
    
    def set_roll_number(self, roll_number):
        self.__roll_number = roll_number
    def display_info(self):
        print(f"Name: {self.__name}, Age: {self.__age}, Roll Number: {self.__roll_number}")
    def update_details(self, name=None, age=None, roll_number=None):
        if name:
            self.set_name(name)
        if age:
            self.set_age(age)
        if roll_number:
            self.set_roll_number(roll_number)

student1 = Student("Alice", 20, "S123")
student1.display_info()
student1.update_details(name="Alice Johnson", age=21)
student1.display_info()


Name: Alice, Age: 20, Roll Number: S123
Name: Alice Johnson, Age: 21, Roll Number: S123


3.Develop a Python program for managing library resources efficiently. Design a class named `LibraryBook`
with attributes like book name, author, and availability status. Implement methods for borrowing and
returning books while ensuring proper encapsulation of attributes.


Tasks
3 1. Create the `LibraryBook` class with encapsulated attributes
3 2. Implement methods for borrowing and returning books
3 3. Ensure proper encapsulation to protect book details
3 4. Test the borrowing and returning functionality with sample data.

In [3]:
class LibraryBook:
    def __init__(self, book_name, author):
        self.__book_name = book_name
        self.__author = author
        self.__is_available = True

    # Method to borrow the book
    def borrow_book(self):
        if self.__is_available:
            self.__is_available = False
            print(f"You have successfully borrowed '{self.__book_name}' by {self.__author}.")
        else:
            print(f"Sorry, '{self.__book_name}' by {self.__author} is currently not available.")

    # Method to return the book
    def return_book(self):
        if not self.__is_available:
            self.__is_available = True
            print(f"Thank you for returning '{self.__book_name}' by {self.__author}.")
        else:
            print(f"'{self.__book_name}' by {self.__author} was not borrowed.")

    # Method to check availability
    def is_book_available(self):
        return self.__is_available

    # Method to get book details
    def get_book_details(self):
        return {
            'book_name': self.__book_name,
            'author': self.__author,
            'is_available': self.__is_available
        }

# Testing the borrowing and returning functionality with sample data
if __name__ == "__main__":
    # Create a LibraryBook object
    book = LibraryBook("1984", "George Orwell")

    # Check availability and borrow the book
    if book.is_book_available():
        book.borrow_book()

    # Try borrowing the book again
    book.borrow_book()

    # Return the book
    book.return_book()

    # Try returning the book again
    book.return_book()

    # Get book details
    details = book.get_book_details()
    print(f"Book Details: {details}")


You have successfully borrowed '1984' by George Orwell.
Sorry, '1984' by George Orwell is currently not available.
Thank you for returning '1984' by George Orwell.
'1984' by George Orwell was not borrowed.
Book Details: {'book_name': '1984', 'author': 'George Orwell', 'is_available': True}


4.Create a simple banking system using object-oriented concepts in Python. Design classes representing
different types of bank accounts such as savings and checking. Implement methods for deposit, withdraw,
and balance inquiry. Utilize inheritance to manage different account types efficiently.


Tasks
3 1. Define base class(es) for bank accounts with common attributes and methods
3 2. Implement subclasses for specific account types (e.g., SavingsAccount, CheckingAccount)
3 3. Provide methods for deposit, withdraw, and balance inquiry in each subclass
3 4. Test the banking system by creating instances of different account types and performing transactions.

In [5]:
class BankAccount:
    def __init__(self, account_number, account_holder, balance=0.0):
        self.account_number = account_number
        self.account_holder = account_holder
        self.balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.balance += amount
            print(f"Deposited {amount} to {self.account_holder}'s account. New balance: {self.balance}")
        else:
            print("Invalid deposit amount.")

    def withdraw(self, amount):
        if 0 < amount <= self.balance:
            self.balance -= amount
            print(f"Withdrew {amount} from {self.account_holder}'s account. New balance: {self.balance}")
        else:
            print("Invalid withdrawal amount or insufficient funds.")

    def get_balance(self):
        print(f"{self.account_holder}'s account balance: {self.balance}")
        return self.balance


class SavingsAccount(BankAccount):
    def __init__(self, account_number, account_holder, balance=0.0, interest_rate=0.02):
        super().__init__(account_number, account_holder, balance)
        self.interest_rate = interest_rate

    def apply_interest(self):
        interest = self.balance * self.interest_rate
        self.balance += interest
        print(f"Applied interest of {interest} to {self.account_holder}'s account. New balance: {self.balance}")


class CheckingAccount(BankAccount):
    def __init__(self, account_number, account_holder, balance=0.0, overdraft_limit=500.0):
        super().__init__(account_number, account_holder, balance)
        self.overdraft_limit = overdraft_limit

    def withdraw(self, amount):
        if 0 < amount <= self.balance + self.overdraft_limit:
            self.balance -= amount
            print(f"Withdrew {amount} from {self.account_holder}'s account. New balance: {self.balance}")
        else:
            print("Withdrawal amount exceeds overdraft limit.")
if __name__ == "__main__":
    savings = SavingsAccount("001", "John Doe", 1000.0)
    savings.get_balance()
    savings.deposit(500)
    savings.apply_interest()
    savings.withdraw(200)
    savings.get_balance()
    
    checking = CheckingAccount("002", "Jane Doe", 500.0)
    checking.get_balance()
    checking.deposit(300)
    checking.withdraw(1000)  
    checking.get_balance()
    checking.withdraw(1000)  
    checking.get_balance()



John Doe's account balance: 1000.0
Deposited 500 to John Doe's account. New balance: 1500.0
Applied interest of 30.0 to John Doe's account. New balance: 1530.0
Withdrew 200 from John Doe's account. New balance: 1330.0
John Doe's account balance: 1330.0
Jane Doe's account balance: 500.0
Deposited 300 to Jane Doe's account. New balance: 800.0
Withdrew 1000 from Jane Doe's account. New balance: -200.0
Jane Doe's account balance: -200.0
Withdrawal amount exceeds overdraft limit.
Jane Doe's account balance: -200.0


5.Write a Python program that models different animals and their sounds. Design a base class called
`Animal` with a method `make_sound()`. Create subclasses like `Dog` and `Cat` that override the
`make_sound()` method to produce appropriate sounds.


Tasks
3 1. Define the `Animal` class with a method `make_sound()`
3 2. Create subclasses `Dog` and `Cat` that override the `make_sound()` method
3 3. Implement the sound generation logic for each subclass
3 4. Test the program by creating instances of `Dog` and `Cat` and calling the `make_sound()` method.

In [6]:

class Animal:
    def make_sound(self):
        raise NotImplementedError("Subclasses must implement this method")
class Dog(Animal):
    def make_sound(self):
        return "Woof! Woof!"

class Cat(Animal):
    def make_sound(self):
        return "Meow! Meow!"


if __name__ == "__main__":
    dog = Dog()
    cat = Cat()
  # Call the make_sound() method for each instance
    print(f"Dog says: {dog.make_sound()}")
    print(f"Cat says: {cat.make_sound()}")


Dog says: Woof! Woof!
Cat says: Meow! Meow!


6].Write a code for Restaurant Management System Using OOPS3
& Create a MenuItem class that has attributes such as name, description, price, and category
& Implement methods to add a new menu item, update menu item information, and remove a menu item
from the menu
& Use encapsulation to hide the menu item's unique identification number
& Inherit from the MenuItem class to create a FoodItem class and a BeverageItem class, each with their own
specific attributes and methods.

In [7]:
class MenuItem:
    def __init__(self, name, description, price, category, item_id):
        self.name = name
        self.description = description
        self.price = price
        self.category = category
        self.__item_id = item_id

    def update_item(self, name=None, description=None, price=None, category=None):
        if name:
            self.name = name
        if description:
            self.description = description
        if price:
            self.price = price
        if category:
            self.category = category

    def remove_item(self):
        del self

    def get_item_details(self):
        return {
            'name': self.name,
            'description': self.description,
            'price': self.price,
            'category': self.category,
            'item_id': self.__item_id
        }

class FoodItem(MenuItem):
    def __init__(self, name, description, price, category, item_id, is_vegetarian):
        super().__init__(name, description, price, category, item_id)
        self.is_vegetarian = is_vegetarian

class BeverageItem(MenuItem):
    def __init__(self, name, description, price, category, item_id, is_alcoholic):
        super().__init__(name, description, price, category, item_id)
        self.is_alcoholic = is_alcoholic

# Testing the Restaurant Management System
if __name__ == "__main__":
    # Create a FoodItem
    pizza = FoodItem("Pizza", "Delicious cheesy pizza", 12.99, "Main Course", 101, is_vegetarian=False)
    print(pizza.get_item_details())

    # Update the FoodItem
    pizza.update_item(price=14.99)
    print(pizza.get_item_details())

    # Create a BeverageItem
    cola = BeverageItem("Coca Cola", "Chilled soft drink", 1.99, "Drinks", 102, is_alcoholic=False)
    print(cola.get_item_details())


{'name': 'Pizza', 'description': 'Delicious cheesy pizza', 'price': 12.99, 'category': 'Main Course', 'item_id': 101}
{'name': 'Pizza', 'description': 'Delicious cheesy pizza', 'price': 14.99, 'category': 'Main Course', 'item_id': 101}
{'name': 'Coca Cola', 'description': 'Chilled soft drink', 'price': 1.99, 'category': 'Drinks', 'item_id': 102}


7]Write a code for Hotel Management System using OOPS 3
& Create a Room class that has attributes such as room number, room type, rate, and availability (private)
& Implement methods to book a room, check in a guest, and check out a guest
& Use encapsulation to hide the room's unique identification number
& Inherit from the Room class to create a SuiteRoom class and a StandardRoom class, each with their own
specific attributes and methods.

In [9]:
class Room:
    def __init__(self, room_number, room_type, rate, availability=True):
        self.__room_number = room_number
        self.room_type = room_type
        self.rate = rate
        self.__availability = availability

    def book_room(self):
        if self.__availability:
            self.__availability = False
            print(f"Room {self.__room_number} has been booked.")
        else:
            print(f"Room {self.__room_number} is not available.")

    def check_in(self):
        if not self.__availability:
            print(f"Guest has checked into room {self.__room_number}.")
        else:
            print(f"Room {self.__room_number} is not booked yet.")

    def check_out(self):
        if not self.__availability:
            self.__availability = True
            print(f"Guest has checked out of room {self.__room_number}. Room is now available.")
        else:
            print(f"Room {self.__room_number} is already available.")

    def get_room_details(self):
        return {
            'room_number': self.__room_number,
            'room_type': self.room_type,
            'rate': self.rate,
            'availability': self.__availability
        }

class SuiteRoom(Room):
    def __init__(self, room_number, rate, has_lounge):
        super().__init__(room_number, "Suite", rate)
        self.has_lounge = has_lounge

class StandardRoom(Room):
    def __init__(self, room_number, rate):
        super().__init__(room_number, "Standard", rate)

# Testing the Hotel Management System
if __name__ == "__main__":
    # Create a SuiteRoom
    suite = SuiteRoom(101, 250.0, has_lounge=True)
    suite.book_room()
    suite.check_in()
    suite.check_out()
    print(suite.get_room_details())

    # Create a StandardRoom
    standard = StandardRoom(102, 100.0)
    standard.book_room()
    standard.check_in()
    print(standard.get_room_details())


Room 101 has been booked.
Guest has checked into room 101.
Guest has checked out of room 101. Room is now available.
{'room_number': 101, 'room_type': 'Suite', 'rate': 250.0, 'availability': True}
Room 102 has been booked.
Guest has checked into room 102.
{'room_number': 102, 'room_type': 'Standard', 'rate': 100.0, 'availability': False}


8.Write a code for Fitness Club Management System using OOPS3
& Create a Member class that has attributes such as name, age, membership type, and membership status
(private)
& Implement methods to register a new member, renew a membership, and cancel a membership
& Use encapsulation to hide the member's unique identification number
& Inherit from the Member class to create a FamilyMember class and an IndividualMember class, each with
their own specific attributes and methods.

In [10]:
class Member:
    def __init__(self, name, age, membership_type, member_id, membership_status="Active"):
        self.name = name
        self.age = age
        self.membership_type = membership_type
        self.__member_id = member_id
        self.__membership_status = membership_status

    def register_member(self):
        print(f"Member {self.name} has been registered with ID {self.__member_id}.")

    def renew_membership(self):
        self.__membership_status = "Active"
        print(f"Membership for {self.name} has been renewed.")

    def cancel_membership(self):
        self.__membership_status = "Cancelled"
        print(f"Membership for {self.name} has been cancelled.")

    def get_member_details(self):
        return {
            'name': self.name,
            'age': self.age,
            'membership_type': self.membership_type,
            'membership_status': self.__membership_status,
            'member_id': self.__member_id
        }

class FamilyMember(Member):
    def __init__(self, name, age, membership_type, member_id, family_members):
        super().__init__(name, age, membership_type, member_id)
        self.family_members = family_members

class IndividualMember(Member):
    def __init__(self, name, age, membership_type, member_id):
        super().__init__(name, age, membership_type, member_id)

# Testing the Fitness Club Management System
if __name__ == "__main__":
    # Create a FamilyMember
    family_member = FamilyMember("John Doe", 45, "Family", 1001, family_members=4)
    family_member.register_member()
    print(family_member.get_member_details())

    # Create an IndividualMember
    individual_member = IndividualMember("Jane Smith", 28, "Individual", 1002)
    individual_member.register_member()
    individual_member.renew_membership()
    print(individual_member.get_member_details())


Member John Doe has been registered with ID 1001.
{'name': 'John Doe', 'age': 45, 'membership_type': 'Family', 'membership_status': 'Active', 'member_id': 1001}
Member Jane Smith has been registered with ID 1002.
Membership for Jane Smith has been renewed.
{'name': 'Jane Smith', 'age': 28, 'membership_type': 'Individual', 'membership_status': 'Active', 'member_id': 1002}


9.Write a code for Event Management System using OOPS3
& Create an Event class that has attributes such as name, date, time, location, and list of attendees (private)
& Implement methods to create a new event, add or remove attendees, and get the total number of
attendees
& Use encapsulation to hide the event's unique identification number
& Inherit from the Event class to create a PrivateEvent class and a PublicEvent class, each with their own
specific attributes and methods.

In [11]:
class Event:
    def __init__(self, name, date, time, location, event_id):
        self.name = name
        self.date = date
        self.time = time
        self.location = location
        self.__event_id = event_id
        self.__attendees = []

    def add_attendee(self, attendee):
        self.__attendees.append(attendee)
        print(f"{attendee} has been added to the event {self.name}.")

    def remove_attendee(self, attendee):
        if attendee in self.__attendees:
            self.__attendees.remove(attendee)
            print(f"{attendee} has been removed from the event {self.name}.")
        else:
            print(f"{attendee} is not attending the event {self.name}.")

    def get_total_attendees(self):
        return len(self.__attendees)

    def get_event_details(self):
        return {
            'name': self.name,
            'date': self.date,
            'time': self.time,
            'location': self.location,
            'event_id': self.__event_id,
            'total_attendees': self.get_total_attendees()
        }

class PrivateEvent(Event):
    def __init__(self, name, date, time, location, event_id, invitation_only):
        super().__init__(name, date, time, location, event_id)
        self.invitation_only = invitation_only

class PublicEvent(Event):
    def __init__(self, name, date, time, location, event_id, max_capacity):
        super().__init__(name, date, time, location, event_id)
        self.max_capacity = max_capacity

# Testing the Event Management System
if __name__ == "__main__":
    # Create a PrivateEvent
    private_event = PrivateEvent("Private Party", "2024-08-15", "19:00", "Private Hall", 2001, invitation_only=True)
    private_event.add_attendee("Alice")
    private_event.add_attendee("Bob")
    print(private_event.get_event_details())

    # Create a PublicEvent
    public_event = PublicEvent("Concert", "2024-08-20", "18:00", "Stadium", 2002, max_capacity=500)
    public_event.add_attendee("Charlie")
    public_event.add_attendee("Diana")
    print(public_event.get_event_details())


Alice has been added to the event Private Party.
Bob has been added to the event Private Party.
{'name': 'Private Party', 'date': '2024-08-15', 'time': '19:00', 'location': 'Private Hall', 'event_id': 2001, 'total_attendees': 2}
Charlie has been added to the event Concert.
Diana has been added to the event Concert.
{'name': 'Concert', 'date': '2024-08-20', 'time': '18:00', 'location': 'Stadium', 'event_id': 2002, 'total_attendees': 2}


10.Write a code for Airline Reservation System using OOPS3
& Create a Flight class that has attributes such as flight number, departure and arrival airports, departure and
arrival times, and available seats (private)
& Implement methods to book a seat, cancel a reservation, and get the remaining available seats
& Use encapsulation to hide the flight's unique identification number
& Inherit from the Flight class to create a DomesticFlight class and an InternationalFlight class, each with their
own specific attributes and methods.

In [12]:
class Flight:
    def __init__(self, flight_number, departure_airport, arrival_airport, departure_time, arrival_time, available_seats, flight_id):
        self.flight_number = flight_number
        self.departure_airport = departure_airport
        self.arrival_airport = arrival_airport
        self.departure_time = departure_time
        self.arrival_time = arrival_time
        self.__available_seats = available_seats
        self.__flight_id = flight_id

    def book_seat(self, seats=1):
        if seats <= self.__available_seats:
            self.__available_seats -= seats
            print(f"{seats} seat(s) booked on flight {self.flight_number}.")
        else:
            print("Not enough seats available.")

    def cancel_reservation(self, seats=1):
        self.__available_seats += seats
        print(f"{seats} seat(s) cancelled on flight {self.flight_number}.")

    def get_remaining_seats(self):
        return self.__available_seats

    def get_flight_details(self):
        return {
            'flight_number': self.flight_number,
            'departure_airport': self.departure_airport,
            'arrival_airport': self.arrival_airport,
            'departure_time': self.departure_time,
            'arrival_time': self.arrival_time,
            'available_seats': self.__available_seats
        }

class DomesticFlight(Flight):
    def __init__(self, flight_number, departure_airport, arrival_airport, departure_time, arrival_time, available_seats, flight_id, domestic_code):
        super().__init__(flight_number, departure_airport, arrival_airport, departure_time, arrival_time, available_seats, flight_id)
        self.domestic_code = domestic_code

    def get_domestic_details(self):
        return {
            **self.get_flight_details(),
            'domestic_code': self.domestic_code
        }

class InternationalFlight(Flight):
    def __init__(self, flight_number, departure_airport, arrival_airport, departure_time, arrival_time, available_seats, flight_id, country_of_origin, country_of_destination):
        super().__init__(flight_number, departure_airport, arrival_airport, departure_time, arrival_time, available_seats, flight_id)
        self.country_of_origin = country_of_origin
        self.country_of_destination = country_of_destination

    def get_international_details(self):
        return {
            **self.get_flight_details(),
            'country_of_origin': self.country_of_origin,
            'country_of_destination': self.country_of_destination
        }

# Testing the Airline Reservation System
if __name__ == "__main__":
    # Create a DomesticFlight
    domestic_flight = DomesticFlight("DL123", "JFK", "LAX", "2024-08-15 10:00", "2024-08-15 13:00", 150, "D001", domestic_code="US-DL123")
    print(domestic_flight.get_domestic_details())
    domestic_flight.book_seat(2)
    print(f"Remaining seats: {domestic_flight.get_remaining_seats()}")

    # Create an InternationalFlight
    international_flight = InternationalFlight("AF456", "CDG", "JFK", "2024-08-20 16:00", "2024-08-20 19:00", 200, "I001", country_of_origin="France", country_of_destination="USA")
    print(international_flight.get_international_details())
    international_flight.book_seat(3)
    print(f"Remaining seats: {international_flight.get_remaining_seats()}")


{'flight_number': 'DL123', 'departure_airport': 'JFK', 'arrival_airport': 'LAX', 'departure_time': '2024-08-15 10:00', 'arrival_time': '2024-08-15 13:00', 'available_seats': 150, 'domestic_code': 'US-DL123'}
2 seat(s) booked on flight DL123.
Remaining seats: 148
{'flight_number': 'AF456', 'departure_airport': 'CDG', 'arrival_airport': 'JFK', 'departure_time': '2024-08-20 16:00', 'arrival_time': '2024-08-20 19:00', 'available_seats': 200, 'country_of_origin': 'France', 'country_of_destination': 'USA'}
3 seat(s) booked on flight AF456.
Remaining seats: 197


11. Define a Python module named constants.py containing constants like pi and the speed of light.

In [14]:
# constants.py
PI = 3.141592653589793
SPEED_OF_LIGHT = 299792458  # in meters per second
GRAVITATIONAL_CONSTANT = 6.67430e-11  # in m^3 kg^-1 s^-2
PLANCK_CONSTANT = 6.62607015e-34  # in Js


12. Write a Python module named calculator.py containing functions for addition, subtraction,
multiplication, and division.

In [17]:
# calculator.py

def add(a, b):
    """Return the sum of a and b."""
    return a + b

def subtract(a, b):
    """Return the difference of a and b."""
    return a - b

def multiply(a, b):
    """Return the product of a and b."""
    return a * b

def divide(a, b):
    """Return the division of a by b. Raise an error if b is zero."""
    if b == 0:
        raise ValueError("Cannot divide by zero!")
    return a / b
import calculator

print(f"Addition: {calculator.add(10, 5)}")
print(f"Subtraction: {calculator.subtract(10, 5)}")
print(f"Multiplication: {calculator.multiply(10, 5)}")
print(f"Division: {calculator.divide(10, 5)}")

13. Implement a Python package structure for a project named ecommerce, containing modules for product
management and order processing.

In [19]:
# product_management.py

class Product:
    def __init__(self, product_id, name, price, quantity):
        self.product_id = product_id
        self.name = name
        self.price = price
        self.quantity = quantity

    def update_quantity(self, new_quantity):
        self.quantity = new_quantity

    def update_price(self, new_price):
        self.price = new_price

    def get_product_info(self):
        return {
            "product_id": self.product_id,
            "name": self.name,
            "price": self.price,
            "quantity": self.quantity
        }


In [20]:
# order_processing.py

class Order:
    def __init__(self, order_id, products):
        self.order_id = order_id
        self.products = products  # List of Product objects

    def calculate_total(self):
        return sum([product.price * product.quantity for product in self.products])

    def add_product(self, product):
        self.products.append(product)

    def remove_product(self, product_id):
        self.products = [product for product in self.products if product.product_id != product_id]

    def get_order_details(self):
        return {
            "order_id": self.order_id,
            "products": [product.get_product_info() for product in self.products],
            "total": self.calculate_total()
        }


14. Implement a Python module named string_utils.py containing functions for string manipulation, such as reversing and capitalizing strings.

In [21]:
# string_utils.py

def reverse_string(s):
    return s[::-1]

def capitalize_string(s):
    return s.capitalize()

def to_uppercase(s):
    return s.upper()

def to_lowercase(s):
    return s.lower()


import string_utils

print(f"Reversed: {string_utils.reverse_string('hello')}")
print(f"Capitalized: {string_utils.capitalize_string('hello')}")
print(f"Uppercase: {string_utils.to_uppercase('hello')}")
print(f"Lowercase: {string_utils.to_lowercase('HELLO')}")

15. Write a Python module named file_operations.py with functions for reading, writing, and appending data to a file.

In [22]:
# file_operations.py

def read_file(file_path):
    """Read the contents of a file and return them as a string."""
    try:
        with open(file_path, 'r') as file:
            return file.read()
    except FileNotFoundError:
        return "File not found."

def write_file(file_path, data):
    """Write data to a file, overwriting any existing content."""
    with open(file_path, 'w') as file:
        file.write(data)
    print(f"Data written to {file_path}.")

def append_file(file_path, data):
    """Append data to a file."""
    with open(file_path, 'a') as file:
        file.write(data)
    print(f"Data appended to {file_path}.")

In [23]:

import file_operations

# Writing data to a file
file_operations.write_file('example.txt', 'Hello, World!')

# Reading data from a file
content = file_operations.read_file('example.txt')
print(f"File content: {content}")

# Appending data to a file
file_operations.append_file('example.txt', '\nWelcome to the file operations module.')

ModuleNotFoundError: No module named 'file_operations'

16. Write a Python program to create a text file named "employees.txt" and write the details of employees,
including their name, age, and salary, into the file.

In [25]:
# Define the employee details
employees = [
    {"name": "John Doe", "age": 30, "salary": 50000},
    {"name": "Jane Smith", "age": 25, "salary": 60000},
    {"name": "Emily Davis", "age": 35, "salary": 70000}
]

# Create and write to the file
with open("employees.txt", "w") as file:
    for employee in employees:
        file.write(f"Name: {employee['name']}, Age: {employee['age']}, Salary: {employee['salary']}\n")

print("Employee details have been written to employees.txt")

Employee details have been written to employees.txt


17. Develop a Python script that opens an existing text file named "inventory.txt" in read mode and displays the contents of the file line by line

In [ ]:
# Open the file in read mode
with open("inventory.txt", "r") as file:
    # Read and display each line
    for line in file:
        print(line.strip())

18. Create a Python script that reads a text file named "expenses.txt" and calculates the total amount spent
on various expenses listed in the file.


with open("expenses.txt", "r") as file:
    total_expenses = 0
    # Read each line in the file
    for line in file:
        # Split the line to get the expense amount
        parts = line.strip().split()
        if parts:
            # Assuming the amount is the last part of the line
            amount = float(parts[-1])
            total_expenses += amount

print(f"Total amount spent: {total_expenses}")

19. Create a Python program that reads a text file named "paragraph.txt" and counts the occurrences of each word in the paragraph, displaying the results in alphabetical order.

In [ ]:
# 19. Count Word Occurrences in a Paragraph

from collections import Counter
import string

def count_word_occurrences(file_name):
    try:
        with open(file_name, 'r') as file:
            text = file.read().lower()
            text = text.translate(str.maketrans("", "", string.punctuation))
            words = text.split()
            word_count = Counter(words)
            for word in sorted(word_count):
                print(f"{word}: {word_count[word]}")
    except FileNotFoundError:
        print(f"File {file_name} not found.")


count_word_occurrences("paragraph.txt")

20. What do you mean by Measure of Central Tendency and Measures of Dispersion .How it can be
calculated.

ANS}Measure of Central Tendency refers to statistical measures that describe the center or typical value in a dataset. The most common measures of central tendency are:
Mean: The average of all data points.
Median: The middle value when data points are arranged in order.
Mode: The value that appears most frequently in the dataset.

Measures of Dispersion describe the spread or variability of data points around the central tendency. Common measures of dispersion include:
Range: The difference between the maximum and minimum values.
Variance: The average of the squared differences from the mean.
Standard Deviation: The square root of the variance, indicating the average distance of data points from the mean.
Interquartile Range (IQR): The difference between the 75th percentile (Q3) and the 25th percentile (Q1), representing the spread of the middle 50% of data.

In [31]:
# example:
import numpy as np

# Sample data
data = [5, 7, 8, 9, 10, 12, 14, 15]

# Measure of Central Tendency
mean = np.mean(data)
median = np.median(data)
mode = max(set(data), key=data.count)  # Simple mode calculation for this example

# Measures of Dispersion
range_value = np.ptp(data)
variance = np.var(data)
std_dev = np.std(data)
q1, q3 = np.percentile(data, [25, 75])
iqr = q3 - q1

print(f"Mean: {mean}")
print(f"Median: {median}")
print(f"Mode: {mode}")
print(f"Range: {range_value}")
print(f"Variance: {variance}")
print(f"Standard Deviation: {std_dev}")
print(f"IQR: {iqr}")


Mean: 10.0
Median: 9.5
Mode: 5
Range: 10
Variance: 10.5
Standard Deviation: 3.24037034920393
IQR: 4.75


21. What do you mean by skewness.Explain its types.Use graph to show.

**Skewness** measures how much a data distribution leans to one side. Here are its main types:

1. **Symmetrical (Zero Skewness)**: Data is evenly spread around the mean. The left and right sides look the same (e.g., normal distribution).

2. **Positive Skewness (Right-Skewed)**: Data has a longer tail on the right. Most values are on the left, with a few larger values pulling the mean to the right.

3. **Negative Skewness (Left-Skewed)**: Data has a longer tail on the left. Most values are on the right, with a few smaller values pulling the mean to the left.

Here's a simple representation using ASCII:

```
Symmetrical:   |-----|-----| (Even)
Right-Skewed:  |-----|--------- (Tail to the right)
Left-Skewed:   ---------|-----| (Tail to the left)
```

22. Explain PROBABILITY MASS FUNCTION (PMF) and PROBABILITY DENSITY FUNCTION (PDF). and what is the difference between them?

### Probability Mass Function (PMF)
- **Used for**: Discrete random variables (e.g., number of dice rolls).
- **Gives**: Probability of a specific value occurring.
- **Example**: Probability of rolling a 3 on a fair six-sided die: \( P(X=3) = \frac{1}{6} \).

### Probability Density Function (PDF)
- **Used for**: Continuous random variables (e.g., height, weight).
- **Gives**: Probability density, not the exact probability of one value. To find the probability, calculate the area under the curve for a range of values.
- **Example**: Probability of a person's height being between 5.5 and 6 feet.

### Key Difference
- **PMF**: For discrete variables; gives exact probability.
- **PDF**: For continuous variables; gives probability density over a range.



23. What is correlation. Explain its type in details.what are the  methods of determining correlation

Correlation measures the strength and direction of the relationship between two variables. It tells us how closely two variables move together.
Types of Correlation:

    Positive Correlation: When one variable increases, the other variable also increases. Example: Height and weight.

    Negative Correlation: When one variable increases, the other variable decreases. Example: The number of absences from class and exam scores.

    No Correlation: No consistent pattern between the two variables. Example: Shoe size and intelligence.

Methods of Determining Correlation:

    Scatter Plot: A graphical representation to visualize the relationship.
    Karl Pearson's Coefficient of Correlation (Pearson's r): Measures linear relationship; ranges from -1 to +1.
    Spearman's Rank Correlation: Used for ordinal data or non-linear relationships.

24. Calculate Coefficient of Correlation Using Karl Pearson’s Method

In [83]:
import math

# Marks obtained by students
accountancy_marks = [85, 90, 78, 92, 88, 76, 85, 89, 84, 91]
statistics_marks = [78, 88, 74, 90, 84, 70, 80, 85, 79, 87]

# Calculate the means
mean_x = sum(accountancy_marks) / len(accountancy_marks)
mean_y = sum(statistics_marks) / len(statistics_marks)

# Calculate the deviations and their products
deviation_products = [(x - mean_x) * (y - mean_y) for x, y in zip(accountancy_marks, statistics_marks)]
squared_deviation_x = [(x - mean_x) ** 2 for x in accountancy_marks]
squared_deviation_y = [(y - mean_y) ** 2 for y in statistics_marks]

# Sum the values
sum_deviation_products = sum(deviation_products)
sum_squared_deviation_x = sum(squared_deviation_x)
sum_squared_deviation_y = sum(squared_deviation_y)

# Calculate the correlation coefficient
correlation_coefficient = sum_deviation_products / math.sqrt(sum_squared_deviation_x * sum_squared_deviation_y)

print(f"Correlation Coefficient: {correlation_coefficient}")


Correlation Coefficient: 0.9808088067239751


25. Differences Between Correlation and Regression

1.Purpose:

    Correlation measures the strength and direction of the relationship between two variables.
    Regression predicts the value of a dependent variable based on the value of an independent variable.

Directionality:

    Correlation does not imply causation or direction.
    Regression assumes a cause-and-effect relationship.

Representation:

    Correlation uses a single number to represent the relationship.
    Regression uses an equation (line of best fit).

Usage:

    Correlation is used to quantify the strength of a relationship.
    Regression is used for prediction and forecasting.

26. Find the Most Likely Price in Delhi corresponding to the price of Rs. 70 at Agra from the following data:Coefficient of correlation between the prices of the two places +0.8.

given Information:

    Coefficient of correlation r=0.8r=0.8
    Price at Agra (xx) = 70
    Let's assume the mean prices (xˉxˉ and yˉyˉ​) and standard deviations (σxσx​ and σyσy​) are known or have a standard scale for direct use with correlation.

Regression Formula

The most likely price in Delhi (yy) can be estimated using a regression line formula when one variable is known:
y−yˉ=r⋅σyσx⋅(x−xˉ)
y−yˉ​=r⋅σx​σy​​⋅(x−xˉ)
Simplified Approach:

If we assume that the means and standard deviations of the prices are equal or close enough, the regression line can be simplified for a direct relationship. This can sometimes be given as:
y=yˉ+r⋅(x−xˉ)
y=yˉ​+r⋅(x−xˉ)

Since the mean (yˉyˉ​) and mean (xˉxˉ) are not provided, the simplest assumption is that the means of both cities' prices are equal, and we can estimate using:
y=r⋅x
y=r⋅x
Calculation:

    r=0.8r=0.8
    x=70x=70

y=0.8⋅70=56
y=0.8⋅70=56
Most Likely Price at Delhi:

The most likely price at Delhi, corresponding to a price of Rs. 70 at Agra, is Rs. 56.



28. What is  Normal Distribution? What are the four Assumptions of Normal Distribution? Explain in detail.

### What is Normal Distribution?

- A **Normal Distribution** is a bell-shaped, symmetric probability distribution.
- The mean, median, and mode are all the same and located at the center.
- It is defined by its mean (µ) and standard deviation (σ).

### Four Assumptions of Normal Distribution

1. **Symmetry**: The distribution is symmetric around the mean. Left and right sides are mirror images.
2. **Unimodal**: It has a single peak, indicating one mode where most data points are concentrated.
3. **Asymptotic**: The tails extend indefinitely without touching the horizontal axis, meaning extreme values are possible but rare.
4. **Empirical Rule (68-95-99.7 Rule)**:
   - 68% of data within ±1 standard deviation of the mean.
   - 95% of data within ±2 standard deviations.
   - 99.7% of data within ±3 standard deviations.

These properties make the normal distribution foundational in statistics for analyzing data and drawing conclusions.

29.Write all the characteristics or Properties of the Normal Distribution Curve



The normal distribution, also known as the Gaussian distribution, has several key characteristics and properties:
1. Symmetry

    The normal distribution curve is perfectly symmetrical around its mean. This means the left side of the curve is a mirror image of the right side¹.

2. Unimodal

    The curve has a single peak, or mode, which occurs at the mean of the distribution. This is why it is called unimodal¹.

3. Mean, Median, and Mode

    In a normal distribution, the mean, median, and mode are all equal and located at the center of the distribution.

4. Asymptotic

    The tails of the normal distribution curve approach, but never touch, the horizontal axis. This means that the probability of extreme values never actually reaches zero.

5. Bell-Shaped Curve

    The shape of the normal distribution curve is bell-shaped, with most of the data points clustering around the mean and fewer data points appearing as you move away from the mean.

6. Defined by Mean and Standard Deviation

    The normal distribution is completely described by its mean (μ) and standard deviation (σ). The mean determines the location of the center of the graph, and the standard deviation determines the height and width of the graph.

7. Area Under the Curve

    The total area under the normal distribution curve is equal to 1. This represents the total probability of all possible outcomes.

8. Empirical Rule (68-95-99.7 Rule)

    Approximately 68% of the data falls within one standard deviation of the mean, 95% falls within two standard deviations, and 99.7% falls within three standard deviations.



30.Which of the following options are  correct about Normal Distribution Curve.
(a) Within a range 0.6745 of σ on both sides the middle 50% of the observations occur i,e. mean ±0.6745σ covers 50% area 25% on each side.
(b) Mean ±1S.D. (i,e.µ ± 1σ) covers 68.268% area, 34.134 % area lies on either side of the mean.
(c) Mean ±2S.D. (i,e. µ ± 2σ) covers 95.45% area, 47.725% area lies on either side of the mean.
(d) Mean ±3 S.D. (i,e. µ ±3σ) covers 99.73% area, 49.856% area lies on the either side of the mean.
(e) Only 0.27% area is outside the range µ ±3σ.

Given:

    Mean (μ) = 60
    Standard Deviation (σ) = 10

Z-Score Formula:
Z=X−μ/σ
